Extension/src/Debugger/configurationProvider.ts

```diff
@@ -7,11 +7,73 @@ import * as debugUtils from './utils';
 import * as os from 'os';
 import * as path from 'path';
 import * as vscode from 'vscode';
+import { getBuildTasks, BuildTaskDefinition } from '../LanguageServer/extension';
+import * as util from '../common';
+import * as fs from 'fs';
+import * as assert from 'assert';
+import * as Telemetry from '../telemetry';
 
 import { IConfiguration, IConfigurationSnippet, DebuggerType, MIConfigurations, WindowsConfigurations, WSLConfigurations, PipeTransportConfigurations } from './configurations';
 import { parse } from 'jsonc-parser';
+import { PlatformInformation } from '../platform';
+
+/*
+ * Retrieves configurations from a provider and displays them in a quickpick menu to be selected.
+ * Ensures that the selected configuration's preLaunchTask (if existent) is populated in the user's task.json.
+ * Automatically starts debugging for "Build and Debug" configurations.
+ */
+export class QuickPickConfigurationProvider implements vscode.DebugConfigurationProvider {
+    private underlyingProvider: vscode.DebugConfigurationProvider;
+
+    public constructor(provider: CppConfigurationProvider) {
+        this.underlyingProvider = provider;
+    }
+
+    async provideDebugConfigurations(folder: vscode.WorkspaceFolder | undefined, token?: vscode.CancellationToken): Promise<vscode.DebugConfiguration[]> {
+        const configs: vscode.DebugConfiguration[] = await this.underlyingProvider.provideDebugConfigurations(folder, token);
+        
+        const editor: vscode.TextEditor = vscode.window.activeTextEditor;
+        if (!editor || !util.fileIsCOrCppSource(editor.document.fileName) || configs.length <= 1) {
+            const defaultConfig: vscode.DebugConfiguration = configs.find(config => { return config.name === "(gdb) Launch"; });
+            assert(defaultConfig);
+            return [defaultConfig];
+        }
+        interface MenuItem extends vscode.QuickPickItem {
+            configuration: vscode.DebugConfiguration;
+        }
 
-abstract class CppConfigurationProvider implements vscode.DebugConfigurationProvider {
+        const items: MenuItem[] = configs.map<MenuItem>(config => {
+            let label: string = config.name;
+            // Rename the menu item for the default configuration as its name is non-descriptive.
+            if (label.indexOf("(gdb) Launch") !== -1) {
+                label = "Default Configuration";
+            }
+            return {label: label, configuration: config};
+        });
+
+        return vscode.window.showQuickPick(items, {placeHolder: "Select a configuration"}).then(async selection => {
+            // Wrap in new Promise to make sure task kicks off before VS Code switches the active document to launch.json
+            return new Promise<vscode.DebugConfiguration[]>(async resolve => {
+                if (selection.label.indexOf("Build and Debug active file") !== -1 && selection.configuration.preLaunchTask) {
```

Comment:
need to handle undefined `selection` (e.g. user cancelled operation)

Reply:
Do you know how to repro this scenario? If not, I can investigate later...

Reply:
I handled it by using the "Default Configuration"...I could cause it to fail with an error too.